hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/DeletedBlockLogImplV2.java

```diff
@@ -0,0 +1,412 @@
+ * memorize the position that scanned by last time, and uses this to
+ * determine where the next scan starts. It has no notion about weight
+ * of each transaction so as long as transaction is still valid, they get
+ * equally same chance to be retrieved which only depends on the nature
+ * order of the transaction ID.
+ */
+public class DeletedBlockLogImplV2
+    implements DeletedBlockLog, EventHandler<DeleteBlockStatus> {
+
+  public static final Logger LOG =
+      LoggerFactory.getLogger(DeletedBlockLogImpl.class);
+
+  private final int maxRetry;
+  private final ContainerManagerV2 containerManager;
+  private final Lock lock;
+  // Maps txId to set of DNs which are successful in committing the transaction
+  private Map<Long, Set<UUID>> transactionToDNsCommitMap;
+  private final DeletedBlockLogStateManager deletedBlockLogStateManager;
+
+  public DeletedBlockLogImplV2(ConfigurationSource conf,
+      ContainerManagerV2 containerManager,
+      SCMRatisServer ratisServer,
+      Table<Long, DeletedBlocksTransaction> deletedTable,
+      BatchOperationHandler batchOperationHandler) {
+    maxRetry = conf.getInt(OZONE_SCM_BLOCK_DELETION_MAX_RETRY,
+        OZONE_SCM_BLOCK_DELETION_MAX_RETRY_DEFAULT);
+    this.containerManager = containerManager;
+    this.lock = new ReentrantLock();
+
+    // transactionToDNsCommitMap is updated only when
+    // transaction is added to the log and when it is removed.
+
+    // maps transaction to dns which have committed it.
+    transactionToDNsCommitMap = new ConcurrentHashMap<>();
+    this.deletedBlockLogStateManager = DeletedBlockLogStateManagerImpl
+        .newBuilder()
+        .setConfiguration(conf)
+        .setDeletedBlocksTable(deletedTable)
+        .setRatisServer(ratisServer)
+        .setBatchOperationHandler(batchOperationHandler)
+        .build();
+  }
+
+
+  @Override
+  public List<DeletedBlocksTransaction> getFailedTransactions()
+      throws IOException {
+    lock.lock();
+    try {
+      final List<DeletedBlocksTransaction> failedTXs = Lists.newArrayList();
+      try (TableIterator<Long,
+          ? extends Table.KeyValue<Long, DeletedBlocksTransaction>> iter =
+               deletedBlockLogStateManager.getReadOnlyIterator()) {
+        while (iter.hasNext()) {
+          DeletedBlocksTransaction delTX = iter.next().getValue();
+          if (delTX.getCount() == -1) {
+            failedTXs.add(delTX);
+          }
+        }
+      }
+      return failedTXs;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @param txIDs - transaction ID.
+   * @throws IOException
+   */
+  @Override
+  public void incrementCount(List<Long> txIDs) throws IOException {
+    lock.lock();
+    try {
+      DeletedBlocksTransactionIDs transactionIDs =
+          DeletedBlocksTransactionIDs.newBuilder().addAllTxID(txIDs).build();
+      deletedBlockLogStateManager
+          .increaseRetryCountOfTransactionDB(transactionIDs);
```

Comment:
Do we really need to propagate the retry count via Ratis? Let's double check on this..

Reply:
Agreed.

I will send an email to master to discuss this idea.

Reply:
In fact, there is a comment to explain the reason of persisting this retry count:

```
        // if the retry time exceeds the maxRetry value
        // then set the retry value to -1, stop retrying, admins can
        // analyze those blocks and purge them manually by SCMCli.
```

So the purpose actually makes sense. 

I think there is optimization room though. For example, only write to DB when the retry count becomes the maximum limit.

Reply:
I sent an email to dev@ozone to discuss this count anyway.

Reply:
Per the discussion in dev@, we can reduce the frequency of writing this count into DB. E.g. write once in every 100 retries. Created a JIRA for it: https://issues.apache.org/jira/browse/HDDS-4708

I think we can continue what we have here. Later we can apply HDDS-4708 to SCM HA to reduce the load on Ratis.